### ISSUE TITLE
3.x, trying to use custom strategy not "getting it".

### ISSUE THREAD

**Issue by alturic** [thephpleague/route#225]
TL;DR: How can I edit all of the exceptions $message property when they are thrown?

---

So, I'm trying to, honestly, simply override the message returned with each of the exceptions. I know I could do something nasty like edit the package directly but obviously that's never going to be proper.

What I did *try* doing was implement a CustomJsonStrategy, and copy the JsonStrategy, and all of the exceptions and Exception itself over verbatim (updating the namespaces of course) but in my custom strategy updated the namespaces for these copied files.

However, it's not that simple as my CustomJsonStrategy complains that `Declaration must be compatible with StrategyInterface->getNotFoundDecorator(exception : \League\Route\Http\Exception\NotFoundException)`

```
public function getNotFoundDecorator(NotFoundException $exception)
    {
        return function (ServerRequestInterface $request, ResponseInterface $response) use ($exception) {
            return $exception->buildJsonResponse($response);
        };
    }
```

Now, I know what I did (for what I was trying to do, simply change the $message for each of the exceptions), is probably nowhere near correct, but I *am* still trying to understand why exactly it doesn't work? I copied the contents of JsonResponse, Exception.php and all of the Exception/*.php files, (updating the namespacing correctly to point to these files) but it doesn't seem to work that way? Primarily confused as to why JsonStrategy works perfect (duh), but my copy/paste/update namespace's throws the errors? I imagine it's due to what the $route->setStrategy() method actually "does" but curious none-the-less.

**Comment by philipobenito** [comment]
To better answer your question it might be better if you explain why you want to change the messages or what the ultimate goal of changing them is?

**Comment by philipobenito** [comment]
I’ll preface this with the fact that I do not recommend what you’re planning to do, it adds a layer of complexity that heavily outweighs any benefit, as any client shouldn’t care about the message.

That being said, it is doable with a custom strategy, I’m not at a computer for a few days though so will come back to you with an example when I am.

**Comment by alturic** [comment]
Let me ask this, and this is primarily due to reading other various questions (funnily enough, one revolving around 404's) where the answer was "Because this isn't a framework, it makes no assumptions about how the user should handle the 404.", how would you "intercept" the packages internal NotFoundException and use one of your own? I ask that with the logic being, using the packages included exceptions, they use 'status_code' for the status code. Almost, forcing you to use the same naming convention for your entire app then.

Again, not being facetious (I love the package), nor being overly obtuse or simplistic, but just trying to understand.

**Comment by philipobenito** [comment]
ApplicationStrategy exceptions just bubble out so you can catch those, JsonStrategy builds a response so not as simple as that but you’d extend the JsonStrategy and overload the decorator methods.

As for the body of a 404 response, it’s there for trivial reasons, you’d get the response code from the response status code itself, not the representation of that in the body.
